src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -810,182 +759,195 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
+
+            annotateLoci(ord, finalType, inserter, product = product)
+
+          case Array(TInterval) =>
+            val intRDD = keyedRDD
+              .map { case (k, v) => (k.getAs[Interval[Locus]](0), v: Annotation) }
+              .sortBy(_._1.start)
+            val partitioner = rdd.orderedPartitioner
+            val overlaps = intRDD.mapPartitionsWithIndex { case (i, it) =>
+              val s = mutable.Set.empty[Int]
+
+              it.foreach { case (interval, _) =>
+                s ++= (partitioner.getPartitionT(interval.start) to partitioner.getPartitionT(interval.end))
+              }
+
+              s.iterator.map(index => (i, index))
+            }.collect()
+              .groupBy(_._2).mapValues(_.map(_._1).toArray)
+
+            type X = (Interval[Locus], Annotation)
+            val adjRDD = new AdjustedPartitionsRDD[X](intRDD,
+              (0 until partitioner.numPartitions).map { i =>
+                overlaps.getOrElse(i, Array.empty[Int])
+                  .map { oldPartIndex => Adjustment[X](oldPartIndex, identity) }
+              })
+
+            val res = rdd.zipPartitions(adjRDD, preservesPartitioning = true) { case (it, intervals) =>
+
+
+              val iTree = IntervalTree.annotationTree[Locus, Annotation](intervals.toArray)
+
+              it.map { case (v, (va, gs)) =>
+                val queries = iTree.queryValues(v.locus)
+                val annot = if (product)
+                  queries: IndexedSeq[Annotation]
+                else
+                  queries.headOption.orNull
+
+                (v, (inserter(va, annot), gs))
+              }
+            }.asOrderedRDD
+
+            copy(rdd = res, vaSignature = finalType)
+          case other =>
+            fatal(
+              s"""method 'annotate_variants_table' expects a key table keyed by one of the following:
+                 |  [ Variant ]
+                 |  [ Locus ]
+                 |  [ Interval ]
+                 |  Found key [ ${ keyTypes.mkString(", ") } ] instead.""".stripMargin)
+        }
     }
+  }
 
-    val (finalType, inserter): (Type, (Annotation, Annotation) => Annotation) =
-      if (isCode) {
-        val ec = EvalContext(Map(
-          "va" -> (0, vaSignature),
-          "table" -> (1, struct)))
-        Annotation.buildInserter(annotationExpr, vaSignature, ec, Annotation.VARIANT_HEAD)
-      } else
-        insertVA(struct, Parser.parseAnnotationRoot(annotationExpr, Annotation.VARIANT_HEAD))
+  def annotateLoci(lociRDD: OrderedRDD[Locus, Locus, Annotation], newSignature: Type,
+    inserter: Inserter, product: Boolean): VariantSampleMatrix[T] = {
 
-    val variantQuery = struct.parseInStructScope[Variant](variantExpr)
+    import LocusImplicits.orderedKey
 
-    val keyedRDD = variantRDD.map {
-      _.map { a =>
-        (variantQuery(a), a: Annotation)
-      }.value
-    }
-      .filter { case (v, a) => v != null }
-      .toOrderedRDD(rdd.orderedPartitioner)
+    val newRDD = if (product)
```

Comment:
This is a bit messy with duplicated logic.  I'm struggling with similar stuff elsewhere.  Here's a potential alternative to consider (compiles, haven't tested):

```
  def annotateLoci(lociRDD: OrderedRDD[Locus, Locus, Annotation], newSignature: Type,
    inserter: Inserter, product: Boolean): VariantSampleMatrix[T] = {

    import LocusImplicits.orderedKey

    def annotate[S](rdd: RDD[(Locus, ((Variant, (Annotation, Iterable[T])), S))],
      ins: (Annotation, S) => Annotation): OrderedRDD[Locus, Variant, (Annotation, Iterable[T])] = {
      rdd
        .mapPartitions({
          it =>
            it.map {
              case (l, ((v, (va, gs)), annotation)) => (v, (ins(va, annotation), gs))
            }
        }, preservesPartitioning = true)
        .asOrderedRDD
    }
    
    val locusKeyedRDD = rdd
      .mapMonotonic(OrderedKeyFunction(_.locus), {
        case (v, vags) => (v, vags)
      })
    
    val newRDD =
      if (product)
      annotate[Array[Annotation]](locusKeyedRDD.orderedLeftJoin(lociRDD),
        (va, a) => inserter(va, a: IndexedSeq[_]))
    else
      annotate[Option[Annotation]](locusKeyedRDD.orderedLeftJoinDistinct(lociRDD),
        (va, a) => inserter(va, a.orNull))
    
    copy(rdd = newRDD, vaSignature = newSignature)
  }
```